orttraining/orttraining/python/training/utils/data/sampler.py

```diff
@@ -0,0 +1,340 @@
+# https://github.com/BaguaSys/bagua/blob/01874a7c3f90904c37c5612a9db866b5d4b8b5ed/bagua/torch_api/contrib/load_balancing_data_loader.py#L12
+class LoadBalancingDistributedSampler:
+    r"""Sampler that balances the data load across workers based on the sample's complexity.
+    This sampler uses a :attr:`complexity_fn` to calculate each sample's computational
+    complexity and make each batch get similar computational complexity.
+    This is useful in scenarios like speech and NLP, where each batch has variable
+    length and distributed training suffers from straggler problem.
+    The usage is similar to `torch.utils.data.DistributedSampler <https://pytorch.org/docs/stable/data.html?highlight=distributedsampler#torch.utils.data.distributed.DistributedSampler>`_,
+    where each process loads a subset of the original dataset that is exclusive to it.
+    .. note::
+        Dataset is assumed to be of constant size (map-style dataset).
+    Args:
+        dataset: Dataset (map-style) used for sampling.
+        complexity_fn(Callable): A function whose input is a sample and output is an integer as a
+            measure of the computational complexity of the sample.
+        world_size (int, optional): Number of processes participating in
+            distributed training. By default, :attr:`world_size` is retrieved from the
+            current distributed group.
+        rank (int, optional): Rank of the current process within :attr:`world_size`.
+            By default, :attr:`rank` is retrieved from the current distributed
+            group.
+        shuffle (bool, optional): If ``True`` (default), sampler will shuffle the
+            indices within the dataset if :attr:`group_size` is None, else will
+            shuffle the groups if :attr:`group_size` is not None.
+        group_size (int, optional): If provided, the dataset will be broken down into
+            :attr:`group_size` sized groups. Indices will only be sorted within the groups
+            and not across the entire dataset. If :attr:`shuffle` is ```True``` and
+            :attr:`group_size` is not ```None```, the position of each group in the dataset
+            will be shuffled. Default: ```None```
+        seed (int, optional): random seed used to shuffle the sampler if
+            :attr:`shuffle=True`. This number should be identical across all
+            processes in the distributed group. Default: 0.
+        drop_last (bool, optional): if ``True``, then the sampler will drop the
+            tail of the data to make it evenly divisible across the number of
+            shards. If ``False``, the sampler will add extra indices to make
+            the data evenly divisible across the shards. Default: ``False``.
+        random_level (float, optional): A float varies from 0 and 1 that controls the extent
+            of load balance. 0 means the best load balance, while 1 means the opposite.
```

Comment:
opposite or just more random? From what I understood, it would be random if random_level is high (>> 1).